Make pre-built functions locale awareness

cosmicshuai opened the pull request:
closes: #4007 

In this PR, some pre-built functions related to format a number, datetime, timespan to a string, or format a string will depend on the locale.
The locale defined in the API will have the highest priority. Then the locale stored in 'turn.activity.locale' will have the second  highest priority. If no locale was specified, the default will be CultureInfo.InvariantCulture. 
Functions support locale awareness:
string
formatNumber
addToTime
startOfDay
startOfHour
startOfMonth
convertToUTC
convertFromUTC
utcNow
getPastTime
getFutureTime
subtractFromTime
formatEpoch
formatTicks
formatDateTime
addDays
toUpper
toLower
sentenceCase
titleCase

xieofxie commented:
Will we make injection locale awareness? I create an example here https://github.com/xieofxie/botbuilder-dotnet/commit/d145177be6feb8d0bfa57d119b959b4eb6930364

cosmicshuai commented:
> Will we make injection locale awareness? I create an example here https://github.com/xieofxie/botbuilder-dotnet/commit/d145177be6feb8d0bfa57d119b959b4eb6930364

I will test it. I think it should work since the locale is stored in the context.

Danieladu commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -2168,6 +2228,38 @@ private static (TimexProperty, string) ParseTimexProperty(object timexExpr)
+            string error = null;
+            if (args.Count == maxArgsLength)
+            {
+                format = args[maxArgsLength - 2] as string;
```
if `maxArgsLength` < 2? < 1?

cosmicshuai replied:
add a condition of maxArgsLength

Danieladu commented:
Please list the functions that support locale, and it is better to have a table to show the difference

cosmicshuai commented:
> Will we make injection locale awareness? I create an example here [xieofxie@d145177](https://github.com/xieofxie/botbuilder-dotnet/commit/d145177be6feb8d0bfa57d119b959b4eb6930364)

After I checked your sample. I think the injected template with locale is not included in current  PR.
We will discuss later to find out a way to incorporate locale in LG injection.

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -610,6 +635,32 @@ public static EvaluateExpressionDelegate ApplyWithError(Func<IReadOnlyList<objec
                 return (value, error);
             };
 
+        public static EvaluateExpressionDelegate ApplyWithErrorAndLocale(Func<IReadOnlyList<object>, CultureInfo, (object, string)> function, VerifyExpression verify = null)
```
>         [](start = 0, length = 8)

Indentation

cosmicshuai replied:
fixed

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -784,12 +838,18 @@ public static ExpressionEvaluator TimeTransform(string type, Func<DateTime, int,
                         {
-                            var formatString = (args.Count() == 3 && args[2] is string string1) ? string1 : DefaultDateTimeFormat;
-                            (value, error) = NormalizeToDateTime(args[0], dt => function(dt, Convert.ToInt32(args[1])).ToString(formatString));
+                            (value, error) = NormalizeToDateTime(args[0], dt => function(dt, Convert.ToInt32(args[1])).ToString(formatString, locale));
```
>locale [](start = 142, length = 6)

Is null allowed here?

cosmicshuai replied:
I don't think the locale here will be null due to DeterminLocale and DetermineFormatAndLocale method will take the CultureInfo.InvariantCulture as default setting.
After set the return locale value in DeterminFormatAndLocale to null. The ToString() method still works. It will return a result as the same as the Thread.CurrentThread.CurrentCulture.

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -784,12 +838,18 @@ public static ExpressionEvaluator TimeTransform(string type, Func<DateTime, int,
                     string error = null;
                     IReadOnlyList<object> args;
                     (args, error) = EvaluateChildren(expr, state, options);
+                    var locale = DetermineLocaleInfo(state);
```
> DetermineLocaleInfo [](start = 32, length = 20)

This is fine as a default, but I should be able to explicitly pass in the locale in all of these places.

cosmicshuai replied:
I see. I change the order of evaluating the locale to ensure that error messages are arranged in the right order. If a locale string was passed explicitly, then it won't try to access the locale stored in the memory.
If no locale given, and locale stored in memory is illegal, it will throw an error.
If no locale given and no locale info stored in memory, take the default locale.

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -3183,34 +3297,68 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
+                                    catch
                                     {
-                                        return args[0].ToString().ToLowerInvariant();
+                                        error = $"the second argument {args[1]} should be should be a locale string.";
```
>th [](start = 50, length = 2)

Initial capital in all of these.

cosmicshuai replied:
fixed

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -3183,34 +3297,68 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
+                                    catch
                                     {
-                                        return args[0].ToString().ToLowerInvariant();
+                                        error = $"the second argument {args[1]} should be should be a locale string.";
```
>should be  [](start = 80, length = 10)

Remove

cosmicshuai replied:
removed

chrimc62 commented:
>                                 error = $"{expr} should contain an ISO format timestamp, a time interval integer, a string unit of time and an optional output format string.";

Update to include locale.

---
Refers to: libraries/AdaptiveExpressions/ExpressionFunctions.cs:3756 in e30a0ea. [](commit_id = e30a0eaf19f339897b0dd882c87a7c6ff189b473, deletion_comment = False)

chrimc62 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -4265,13 +4528,54 @@ private static IEnumerable<object> Flatten(IEnumerable<object> list, int dept)
+                            return (result, error);
+                        }, 
+                    ReturnType.String,
+                    expr => ValidateOrder(expr, new[] { ReturnType.String }, ReturnType.Object)),
```
>expr => ValidateOrder(expr, new[] { ReturnType.String }, ReturnType.Object)) [](start = 20, length = 76)

Not sure I get the semantics here.  I should be able to do string('foo') or string('foo', 'en-us') or string(3000, 'fr-fr').  Where is the optional thing an object?

cosmicshuai replied:
for ValidateOrder method, the signature is expression, optional parameters array, fixed parameters. 
So new[] { ReturnType.String } (optional) comes before ReturnType.Object

chrimc62 commented on tests/AdaptiveExpressions.Tests/BadExpressionTests.cs:
```diff
@@ -259,34 +262,34 @@ public class BadExpressionTests
             Test("getPastTime(timestamp, 'W')"), // error parameters format
             Test("getPastTime('yyyy', '1')"), // second param should be integer
-            Test("getPastTime('yyyy')"), // should have 2 or 3 params
+            Test("getPastTime('yyyy')"), // should have 2 or 3 or 4params
```
>4params [](start = 66, length = 7)

space

cosmicshuai replied:
fixed

cosmicshuai commented on tests/AdaptiveExpressions.Tests/BadExpressionTests.cs:
```diff
@@ -259,34 +262,34 @@ public class BadExpressionTests
             Test("getFutureTime('yyyy', '1')"), // second param should be integer
-            Test("getFutureTime('yyyy')"), // should have 2 or 3 params
+            Test("getFutureTime('yyyy')"), // should have 2 or 3 or 4 params
+            Test("utcNow(hello, hello, hellow)"), // should have 0 or 1 or 2 params
```
more bad expressions tests are added and comments updated, some parameters change to legal locale strings

cosmicshuai commented:
> > ```
> >                             error = $"{expr} should contain an ISO format timestamp, a time interval integer, a string unit of time and an optional output format string.";
> > ```
> 
> Update to include locale.
> 
> Refers to: libraries/AdaptiveExpressions/ExpressionFunctions.cs:3756 in [e30a0ea](https://github.com/microsoft/botbuilder-dotnet/commit/e30a0eaf19f339897b0dd882c87a7c6ff189b473). [](commit_id = [e30a0ea](https://github.com/microsoft/botbuilder-dotnet/commit/e30a0eaf19f339897b0dd882c87a7c6ff189b473), deletion_comment = False)

related error messages are updated

cleemullins commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -548,6 +558,16 @@ public static (IReadOnlyList<object>, string error) EvaluateChildren(Expression
 
         // Apply -- these are helpers for adding functions to the expression library.
 
+        public static string EvaluateLocale(IMemory state)
```
Public method. Public class. Comments required.

cosmicshuai replied:
changed to private

cleemullins commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -2168,6 +2224,60 @@ private static (TimexProperty, string) ParseTimexProperty(object timexExpr)
+            {
+                result = new CultureInfo(localeStr);
+            }
+            catch
```
Consider typing this exception.

cosmicshuai replied:
changed to the e.message as the final error.

tomlm commented:
>         public (object value, string error) TryEvaluate(object state, Options options = null)

This needs a new signature which takes Locale as a parameter. 
We should consider strongly setting thread.SetLocaleStorage()


---
Refers to: libraries/AdaptiveExpressions/Expression.cs:492 in 1e94f37. [](commit_id = 1e94f377efa92670fce1a390fa1c6f7fa92683ca, deletion_comment = False)

cosmicshuai commented:
@tomlm 
I added locale as a parameter in tryEvaluate. 
The locale is set as `Thread.SetData(Thread.GetNamedDataSlot("locale"), locale);`
The priority for determining locale is API > Thread.GetData > turn.activity.locale > default(invariant culture).
Some tests based this routine is added.
Do you have any comments or suggestions?

boydc2014 commented:
> > ```
> >     public (object value, string error) TryEvaluate(object state, Options options = null)
> > ```
> 
> This needs a new signature which takes Locale as a parameter.
> We should consider strongly setting thread.SetLocaleStorage()
> 
> Refers to: libraries/AdaptiveExpressions/Expression.cs:492 in [1e94f37](https://github.com/microsoft/botbuilder-dotnet/commit/1e94f377efa92670fce1a390fa1c6f7fa92683ca). [](commit_id = [1e94f37](https://github.com/microsoft/botbuilder-dotnet/commit/1e94f377efa92670fce1a390fa1c6f7fa92683ca), deletion_comment = False)

Do you prefer include locale explicitly in TryEvaluate, or let locale sits inside `options`, i feel the latter is better and less disruptively of the current interface, what do you think?

boydc2014 commented:
> @tomlm
> I added locale as a parameter in tryEvaluate.
> The locale is set as `Thread.SetData(Thread.GetNamedDataSlot("locale"), locale);`
> The priority for determining locale is API > Thread.GetData > turn.activity.locale > default(invariant culture).
> Some tests based this routine is added.
> Do you have any comments or suggestions?

I don't think we should hard-code turn.activity.locale inside expression, we should let upper layer, let's say adapter layer set turn.activity.locale into thread.data, which we (expression) layer pick from. 

hard code turn.activity.locale seems very artificial

boydc2014 commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -72,6 +73,16 @@ public static class ExpressionFunctions
+        /// <summary>
+        /// The Path stores the locale info in memory.
+        /// </summary>
+        private static readonly string LocaleMemoryPath = "turn.activity.locale";
```
I mean, here, this knowledge "turn.activity.locale" should not flow into expression library.

cosmicshuai replied:
moved to ActionScope

boydc2014 commented on libraries/AdaptiveExpressions/Expression.cs:
```diff
@@ -527,10 +530,16 @@ public void ValidateTree()
             var opts = options ?? new Options();
+            if (locale != null)
+            {
+                Thread.SetData(Thread.GetNamedDataSlot("locale"), locale);
```
The other question is for storage should we use Thread.GetNamedDataSlot or we should use thread.currentThread.culture? 

cc @tomlm @chrimc62

cosmicshuai replied:
now stored in Thread.CurrentThread.CurrentCulture

cosmicshuai commented:
@tomlm @chrimc62 @boydc2014 
I moved the way set the CultureInfo to thread. Now it is happened during BeginActionAsync in ActionScope.
And the locale signature in TryEvaluate is put into Options. 
The locale info is stored in Thread.CurrentThread.CurrentCulture.
According to these changes, I modified some test in AdaptiveExpressions. Also I added a E2E test in LGGeneratorTest

boydc2014 commented on libraries/Microsoft.Bot.Builder.LanguageGeneration/Evaluator.cs:
```diff
@@ -22,8 +23,8 @@ namespace Microsoft.Bot.Builder.LanguageGeneration
         public const string LGType = "lgType";
-
         private const string ReExecuteSuffix = "!";
+        private const string LocaleExpression = "turn.activity.locale";
```
we should not hard code any path in LG

cosmicshuai replied:
fixed

boydc2014 commented on libraries/Microsoft.Bot.Builder.Dialogs/DialogContextExtension.cs:
```diff
@@ -0,0 +1,49 @@
+        /// </summary>
+        /// <param name="dialogContext">The dialogContext to extract information.</param>
+        /// <returns>A <see cref="CultureInfo"/> representing the current locale.</returns>
+        public static CultureInfo EvalLocaleInfo(this DialogContext dialogContext)
```
Maybe, call this GetLocale?

boydc2014 replied:
And the return type should be string for better portability.

cosmicshuai replied:
changed

boydc2014 commented on libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:
```diff
@@ -94,6 +97,14 @@ public EvaluationOptions(IList<string> optionsList)
+        /// <value>
+        /// A CultureInfo or null object.
+        /// </value>
+        public CultureInfo Locale { get; set; } = null;
```
string would be good.

cosmicshuai replied:
changed

Danieladu commented on libraries/AdaptiveExpressions/Expression.cs:
```diff
@@ -530,6 +530,18 @@ public void ValidateTree()
             var opts = options ?? new Options();
+            if (opts.Locale != null)
+            {
+                try
```
simplify the empty try catch to -> "new CultureInfo(opts.Locale);"

cosmicshuai replied:
fixed

Danieladu commented on libraries/AdaptiveExpressions/ExpressionFunctions.cs:
```diff
@@ -2184,6 +2261,55 @@ private static (TimexProperty, string) ParseTimexProperty(object timexExpr)
+            {
+                if (args.Count == maxArgsLength)
+                {
+                    format = args[maxArgsLength - 2] as string;
```
add comments here, why "maxArgsLength - 2"?

cosmicshuai replied:
comments added

Danieladu commented on libraries/AdaptiveExpressions/Options.cs:
```diff
@@ -1,4 +1,5 @@
 ﻿using System;
+using System.Globalization;
```
unused

cosmicshuai replied:
removed

Danieladu commented on libraries/Microsoft.Bot.Builder.Dialogs/DialogContextExtension.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// Provides extension methods for <see cref="DialogContext"/>.
+    /// </summary>
+    public static class DialogContextExtension
```
Maybe this class could be merged into the "Extensions" class ?

cosmicshuai replied:
merged into DialogExtensions

Danieladu commented on libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/ActionScope.cs:
```diff
@@ -3,12 +3,11 @@
 
 using System;
 using System.Collections.Generic;
+using System.Globalization;
```
unused change for this file

cosmicshuai replied:
removed

tomlm commented on libraries/Microsoft.Bot.Builder.Dialogs/DialogExtensions.cs:
```diff
@@ -94,6 +99,27 @@ public static async Task RunAsync(this Dialog dialog, ITurnContext turnContext,
+                }
+            }
+
+            return locale;
```
umm... how about just this:
{
   return dc.State.GetValue<string>("turn.locale");
}
 ?

cosmicshuai replied:
added Locale property in TurnContext.
And now the scenario is to set the locale as 
```C#
await CreateFlow(async (turnContext, cancellationToken) =>
            {
                turnContext.Locale = "de-DE";
                await dm.OnTurnAsync(turnContext, cancellationToken: cancellationToken).ConfigureAwait(false);
            })
```

Initially, I used setProperty action to set an locale
```
{
    "$kind": "Microsoft.SetProperty",
    "property": "turn.activity.locale",
    "value": "de-DE"
}

```

boydc2014 replied:
@cosmicshuai @tomlm if we change from turn.activity.locale to turn.locale, where we init turn.locale, in adapter? both init turn.locale and init thread.currentCulture, make sense?

tomlm commented on libraries/Microsoft.Bot.Builder/Adapters/TestAdapter.cs:
```diff
@@ -445,6 +445,7 @@ public Activity MakeActivity(string text = null)
                 Text = text,
             };
 
+            Thread.CurrentThread.CurrentCulture = new CultureInfo(activity.Locale);
```
> Thread.CurrentThread.CurrentCulture = new CultureInfo(activity.Locale); [](start = 11, length = 72)

No this is not the right place to do this.

cosmicshuai replied:
changed

tomlm commented:
>             using (var context = new TurnContext(this, activity))

This is the place to do this:
if (activity.Locale != null)
{
    Thread.CurrentThread.CurrentCulture = new CultureInfo(activity.Locale);
}


---
Refers to: libraries/Microsoft.Bot.Builder/Adapters/TestAdapter.cs:193 in de880c8. [](commit_id = de880c8e9a4bce9d30676a4cf7b6587e60c858a5, deletion_comment = False)

tomlm commented:
>                 await RunPipelineAsync(context, callback, cancellationToken).ConfigureAwait(false);

We should change RunPipelineAsync to do this:
    Thread.CurrentThread.CurrentCulture = new CultureInfo(activity.Locale);

That way it will do it for ALL adapters.

---
Refers to: libraries/Microsoft.Bot.Builder/Adapters/TestAdapter.cs:195 in de880c8. [](commit_id = de880c8e9a4bce9d30676a4cf7b6587e60c858a5, deletion_comment = False)

tomlm commented on libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/TrueSelector.cs:
```diff
@@ -36,6 +39,7 @@ public override Task<IReadOnlyList<OnCondition>> SelectAsync(ActionContext conte
                 {
                     var expression = conditional.GetExpression();
+                    var opt = new Options() { Locale = context.GetLocale() };
                     var (value, error) = expression.TryEvaluate(context.State);
```
>State [](start = 72, length = 5)

you aren't using options.

cosmicshuai replied:
fixed

tomlm commented on libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/Actions/AssertCondition.cs:
```diff
@@ -39,6 +41,7 @@ public AssertCondition([CallerFilePath] string path = "", [CallerLineNumber] int
         {
+            var opt = new Options() { Locale = dc.GetLocale() };
             var (result, error) = Condition.TryEvaluate(dc.State);
             if ((bool)result == false)
```
you arne't using options

cosmicshuai replied:
fixed

tomlm commented on libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:
```diff
@@ -38,13 +38,15 @@ public EvaluationOptions()
             this.StrictMode = opt.StrictMode;
             this.NullSubstitution = opt.NullSubstitution;
             this.LineBreakStyle = opt.LineBreakStyle;
+            this.Locale = opt.Locale;
```
>; [](start = 36, length = 1)

= opt.Locale ?? Thread.CurrentThreadLocale;

cosmicshuai replied:
changed

tomlm commented on libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Actions/SwitchCondition.cs:
```diff
@@ -148,7 +150,8 @@ public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc,
             foreach (var caseScope in this.Cases)
             {
-                var (value, error) = this.caseExpressions[caseScope.Value].TryEvaluate(dc.State);
+                var opt = new Options() { Locale = dc.GetLocale() };
```
If we change the options to default to Thread.CurrentCulture then there is no need to change all of these calls.

cosmicshuai replied:
I changed the default locale to Thread.CurrentThread.CurrentCulture.Name.
It will always use the default locale other the locale stored in the turn.locale.
I have a question for these actions and selectors.
Should the expressions used in these cases are locale independent?

boydc2014 commented:
@cosmicshuai seems like the commit history is now messed up, can you do clean up, a full repush is also acceptable, i guess
